krm-functions/nfdeploy-fn/common/mutator.go

```diff
@@ -0,0 +1,206 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	nephiodeployv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+	nephioreqv1alpha1 "github.com/nephio-project/api/nf_requirements/v1alpha1"
+	infrav1alpha1 "github.com/nephio-project/nephio-controller-poc/apis/infra/v1alpha1"
+	kptcondsdk "github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	"github.com/nephio-project/nephio/krm-functions/lib/kubeobject"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"reflect"
+)
+
+type NfType interface {
+	*nephiodeployv1alpha1.UPFDeployment | *nephiodeployv1alpha1.AMFDeployment | *nephiodeployv1alpha1.SMFDeployment
+}
+
+func Run[T NfType](rl *fn.ResourceList, gvk schema.GroupVersionKind) (bool, error) {
+	nfDeployFn := NewMutator[T](gvk)
+
+	var err error
+
+	kptfile := rl.Items.GetRootKptfile()
+	nfDeployFn.pkgName = kptfile.GetName()
+
+	nfDeployFn.sdk, err = kptcondsdk.New(
+		rl,
+		&kptcondsdk.Config{
+			For: corev1.ObjectReference{
+				APIVersion: nephiodeployv1alpha1.GroupVersion.Identifier(),
+				Kind:       nfDeployFn.gvk.Kind,
+			},
+			Watch: map[corev1.ObjectReference]kptcondsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       reflect.TypeOf(infrav1alpha1.ClusterContext{}).Name(),
+				}: nfDeployFn.ClusterContextCallBackFn,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       reflect.TypeOf(nephioreqv1alpha1.Capacity{}).Name(),
+				}: nfDeployFn.CapacityContextCallBackFn,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       nephioreqv1alpha1.InterfaceKind,
+				}: nfDeployFn.InterfaceCallBackFn,
+				{
+					APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+					Kind:       nephioreqv1alpha1.DataNetworkKind,
+				}: nfDeployFn.DnnCallBackFn,
+			},
+			GenerateResourceFn: nfDeployFn.GenerateResourceFn,
+		},
+	)
+
+	if err != nil {
+		rl.Results.ErrorE(err)
+		return false, err
+	}
+
+	return nfDeployFn.sdk.Run()
+}
+
+func (h *NfDeployFn[T1]) ClusterContextCallBackFn(o *fn.KubeObject) error {
+	var cluster infrav1alpha1.ClusterContext
+	err := o.As(&cluster)
+	if err != nil {
+		return err
+	}
+	h.site = *cluster.Spec.SiteCode
+	return nil
+}
+
+func (h *NfDeployFn[T1]) CapacityContextCallBackFn(o *fn.KubeObject) error {
+	var capacity nephioreqv1alpha1.Capacity
+	err := o.As(&capacity)
+	if err != nil {
+		return err
+	}
+
+	h.capacity = capacity
+	return nil
+}
+
+func (h *NfDeployFn[T1]) InterfaceCallBackFn(o *fn.KubeObject) error {
+	var itfce *nephioreqv1alpha1.Interface
+	err := o.As(&itfce)
```

Comment:
Is this right? itfce is a pointer already.

Reply:
It still works because kubeobject `As` method is implemented to handle pointer.

Reply:
Thanks for pointing out. I've fixed it by moving to kuceobjectExt library.